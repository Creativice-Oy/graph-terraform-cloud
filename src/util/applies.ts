import fetch from 'node-fetch';

/**
 * See: https://www.terraform.io/docs/internals/machine-readable-ui.html#operation-messages
 */
export type ParsedTerraformApplyLogActionType =
  | 'apply_start'
  | 'apply_progress'
  | 'apply_complete'
  | 'apply_errored'
  | 'provision_start'
  | 'provision_progress'
  | 'provision_complete'
  | 'provision_errored'
  | 'refresh_start'
  | 'refresh_complete';

export type ParsedTerraformApplyLogHook = {
  resource: {
    addr: string;
    module: string;
    resource: string;
    implied_provider: string;
    resource_type: string;
    resource_name: string;
    resource_key: string | null;
  };
  action: string;
  id_key: string;
  id_value: string;
  elapsed_seconds: number;
};

export type ParsedTerraformApplyLog = {
  '@level': string;
  '@message': string;
  '@module': string;
  '@timestamp': string;
  type: string;
  hook: ParsedTerraformApplyLogHook;
};

export async function downloadAndParseRunApply(url: string) {
  const response = await fetch(url);
  return response.text();
}

/**
 * When the `Structured Run Output` workspace option is enabled, Terraform Cloud
 * stores raw logs of the Terraform apply.
 */
export function parseRunApply(text: string): ParsedTerraformApplyLog[] {
  const splitApply = text.split('\n');
  const collected: ParsedTerraformApplyLog[] = [];

  for (const row of splitApply) {
    try {
      const parsed = JSON.parse(row) as ParsedTerraformApplyLog;

      // Validate that one of the apply log properties are available before
      // considering it valid
      if (parsed['@message']) {
        collected.push(parsed);
      }
    } catch (err) {
      // This apply log is likely using non-structured run output and therefore
      // cannot be parsed
    }
  }
  return collected;
}

export async function collectApplyCompleteOperations(url: string) {
  const text = await downloadAndParseRunApply(url);
  const parsed = parseRunApply(text);
  return parsed.filter((v) => v.type === 'apply_complete');
}
