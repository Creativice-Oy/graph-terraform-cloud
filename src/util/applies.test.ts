import { promises as fs } from 'fs';
import * as path from 'path';
import { parseRunApply } from './applies';

async function loadFixture(fixtureName: string): Promise<string> {
  const fullPath = path.join(__dirname, '__fixtures__', `${fixtureName}.txt`);
  const file = await fs.readFile(fullPath, { encoding: 'utf-8' });
  return file;
}

async function parseRunApplyFixture(fixtureName: string) {
  return parseRunApply(await loadFixture(fixtureName));
}

describe('#parseRunApply', () => {
  test('should parse "resource-added" log output', async () => {
    expect(await parseRunApplyFixture('resource-added')).toMatchSnapshot();
  });
});
