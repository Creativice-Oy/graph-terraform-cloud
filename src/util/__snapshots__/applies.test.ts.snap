// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#parseRunApply should parse "resource-added" log output 1`] = `
Array [
  Object {
    "@level": "info",
    "@message": "Terraform 1.0.2",
    "@module": "terraform.ui",
    "@timestamp": "2021-07-22T13:15:29.223761Z",
    "terraform": "1.0.2",
    "type": "version",
    "ui": "0.1.0",
  },
  Object {
    "@level": "info",
    "@message": "google_storage_bucket.sink-log-bucket: Plan to create",
    "@module": "terraform.ui",
    "@timestamp": "2021-07-22T13:15:30.355716Z",
    "change": Object {
      "action": "create",
      "resource": Object {
        "addr": "google_storage_bucket.sink-log-bucket",
        "implied_provider": "google",
        "module": "",
        "resource": "google_storage_bucket.sink-log-bucket",
        "resource_key": null,
        "resource_name": "sink-log-bucket",
        "resource_type": "google_storage_bucket",
      },
    },
    "type": "planned_change",
  },
  Object {
    "@level": "info",
    "@message": "google_storage_bucket.sink-log-bucket: Creating...",
    "@module": "terraform.ui",
    "@timestamp": "2021-07-22T13:15:30.898372Z",
    "hook": Object {
      "action": "create",
      "resource": Object {
        "addr": "google_storage_bucket.sink-log-bucket",
        "implied_provider": "google",
        "module": "",
        "resource": "google_storage_bucket.sink-log-bucket",
        "resource_key": null,
        "resource_name": "sink-log-bucket",
        "resource_type": "google_storage_bucket",
      },
    },
    "type": "apply_start",
  },
  Object {
    "@level": "info",
    "@message": "google_storage_bucket.sink-log-bucket: Creation complete after 1s [id=j1-gc-integration-dev-v3-test-tf-bucket]",
    "@module": "terraform.ui",
    "@timestamp": "2021-07-22T13:15:31.754196Z",
    "hook": Object {
      "action": "create",
      "elapsed_seconds": 1,
      "id_key": "id",
      "id_value": "j1-gc-integration-dev-v3-test-tf-bucket",
      "resource": Object {
        "addr": "google_storage_bucket.sink-log-bucket",
        "implied_provider": "google",
        "module": "",
        "resource": "google_storage_bucket.sink-log-bucket",
        "resource_key": null,
        "resource_name": "sink-log-bucket",
        "resource_type": "google_storage_bucket",
      },
    },
    "type": "apply_complete",
  },
  Object {
    "@level": "info",
    "@message": "Apply complete! Resources: 1 added, 0 changed, 0 destroyed.",
    "@module": "terraform.ui",
    "@timestamp": "2021-07-22T13:15:32.063540Z",
    "changes": Object {
      "add": 1,
      "change": 0,
      "operation": "apply",
      "remove": 0,
    },
    "type": "change_summary",
  },
  Object {
    "@level": "info",
    "@message": "Outputs: 0",
    "@module": "terraform.ui",
    "@timestamp": "2021-07-22T13:15:32.063645Z",
    "outputs": Object {},
    "type": "outputs",
  },
]
`;
